checks/permissions.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"errors"
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/ossf/scorecard/checker"
+)
+
+const CheckPermissions = "Token-Permissions"
+
+// ErrInvalidGitHubWorkflowFile : Invalid GitHub workflow file.
+var ErrInvalidGitHubWorkflowFile = errors.New("invalid GitHub workflow file")
+
+//nolint:gochecknoinits
+func init() {
+	registerCheck(CheckPermissions, leastPrivilegedTokens)
+}
+
+func leastPrivilegedTokens(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(CheckPermissions, ".github/workflows/*", true, c, validateGitHubActionTokenPermissions)
+}
+
+func validatePermission(key string, value interface{}, path string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	switch val := value.(type) {
+	case string:
+		if val == "write" {
+			logf("!! token-permissions/github-token - %v permission set to '%v' in %v", key, val, path)
+			return false, nil
+		}
+	default:
+		return false, ErrInvalidGitHubWorkflowFile
+	}
+	return true, nil
+}
+
+func validateReadPermissions(config map[interface{}]interface{}, path string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	permissionFound := false
+	permissionRead := true
+	var r bool
+	var err error
+
+	// Iterate over the values.
+	for key, value := range config {
+		if key != "permissions" {
+			continue
+		}
+
+		// We have found the permissions keyword.
+		permissionFound = true
+
+		// Check the type of our values.
+		switch val := value.(type) {
+		// Empty string is nil type.
+		// It defaults to 'none'
+		case nil:
+
+		// String type.
+		case string:
+			fmt.Println(val)
+			if val != "read-all" && val != "" {
+				logf("!! token-permissions/github-token - permission set to '%v' in %v", val, path)
+				return false, nil
+			}
+
+		// Map type.
+		case map[interface{}]interface{}:
+
+			// Iterate over the permission, verify keys and values are strings.
+			for k, v := range val {
```

Comment:
I saw the comments from @oliverchang  about static binding. 
But this custom marshaling could cause errors in the future and more than that readability is an issue. 

In `go`  `Clear is better than clever.` https://go-proverbs.github.io/  . I would have voted for static binding.

Reply:
while re-writing it, I almost reverted to the old code because it was hard to read. I think readability is OKish.
what do you mean by `custom marshaling could cause errors in the future`?
do you know how to use union with golang yaml parser? That's what I would have liked to use, but I don't know it it's feasible.

Reply:
> while re-writing it, I almost reverted to the old code because it was hard to read. I think readability is OKish.
> what do you mean by `custom marshaling could cause errors in the future`?

we assume that if it is of type `string` then `validatePermission` would work along with that. This is an assumption and there isn't a contract by the provider that it would in the future

> do you know how to use union with golang yaml parser? That's what I would have liked to use, but I don't know it it's feasible. If it is, I could use that instead.

TBH I don't know. I think I would have still keep it simple.

Reply:
> > while re-writing it, I almost reverted to the old code because it was hard to read. I think readability is OKish.
> > what do you mean by `custom marshaling could cause errors in the future`?
> 
> we assume that if it is of type `string` then `validatePermission` would work along with that. This is an assumption and there isn't a contract by the provider that it would in the future

We verify it is a string before calling the function.

Reply:
an alternative approach to handle unions could be to first try to unmarshal statically with permissions defined as a `string`; and if it fails try again with the permission defined as a map (contents, pullrequest, etc) as in the original commit.

WDUT?

Reply:
I would disagree that custom marshalling is less readable. The previous approach resulted in many bits of copy-pasted code which is hard to read through and easy to introduce mistakes in. 

Given that Go doesn't handle unions or dynamically typed structs very well (as in this case), we probably need something like this anyway.

Reply:
Going to merge for now with the existing code. Happy to revisit once I have feedback on the above.